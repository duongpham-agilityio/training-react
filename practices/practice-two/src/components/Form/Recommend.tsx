import { memo, useCallback } from 'react';

// Styles
import styles from 'components/Form/index.module.css';

// Types
import { Recommend } from 'types';

export interface RecommendProps {
  books: Recommend[];
  onSelect: (index: number) => void;
}

const BookRecommend = ({ books, onSelect }: RecommendProps) => {
  const renderBook = useCallback(
    (book: Recommend, index: number) => {
      const { name } = book;
      const handleSelectBook = () => onSelect(index);

      return (
        <li
          className={styles.recommendItem}
          key={index}
          onClick={handleSelectBook}
        >
          <p>{name}</p>
        </li>
      );
    },
    [onSelect]
  );

  return <ul className={styles.recommend}>{books.map(renderBook)}</ul>;
};
export default memo(BookRecommend);
