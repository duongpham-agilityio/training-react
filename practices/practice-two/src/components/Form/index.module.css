.overlay {
  position: fixed;
  inset: 0;
  z-index: 9999;
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: rgba(0, 0, 0, 0.2);
  font-family: var(--primary-font);
}

.form {
  width: 800px;
  height: 624px;
  background-color: white;
  border-radius: 15px;
  padding: 22px;
}

.heading {
  text-align: center;
  padding: 0 0 20px;
}

.content {
  display: flex;
  flex-direction: column;
  gap: 24px;
  padding: 20px 0;
}

.formItem {
  display: flex;
  align-items: start;
  gap: 60px;
}

.formItem .upload {
  position: relative;
  display: flex;
  align-items: start;
  gap: 26px;
}

.uploadItem:first-child {
  flex: 1;
}

.formField {
  flex: 1;
  position: relative;
}

.recommend {
  position: absolute;
  top: calc(100% + 20px);
  border-radius: 7px;
  background-color: white;
  z-index: 999;
  box-shadow: 0 6px 10px silver;
}

.recommendItem {
  padding: 10px 7px;
}

.recommendItem:hover {
  background-color: var(--gray-color);
  cursor: pointer;
}

.text {
  height: 118px;
}

.input {
  background-color: transparent;
  position: absolute;
}

.label {
  display: inline-block;
  font-weight: 700;
  padding: 10px 0;
}

.action {
  display: flex;
  align-items: center;
  justify-content: end;
  gap: 20px;
}

.image {
  width: 95px;
  height: 116px;
  object-fit: cover;
  border-radius: 7px;
}
