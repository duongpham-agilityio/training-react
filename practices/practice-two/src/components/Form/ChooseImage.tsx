import { memo } from 'react';

// Components
import { Button } from 'components/commons';

// Assets
import UploadIcon from 'assets/icons/upload.svg';

const ChooseImage = ({
  name,
  chooseHandler,
}: {
  name: string;
  chooseHandler: () => void;
}) =>
  !name ? (
    <Button
      label="Upload"
      width="w-lg"
      border="b-lg"
      variant="primary"
      leftIcon={UploadIcon}
      onClick={chooseHandler}
      data-testid="upload-btn"
    />
  ) : (
    <p onClick={chooseHandler}>{name}</p>
  );

export default memo(ChooseImage);
