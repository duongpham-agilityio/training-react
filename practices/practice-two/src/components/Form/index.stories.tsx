import { useCallback } from 'react';
import { Meta, Story } from '@storybook/react';

// Components
import Form, { FormProps } from 'components/Form';

// Hooks
import { FormData } from 'hooks';

export const Template: Story<FormProps> = () => {
  const onSubmit = useCallback(async (book: FormData) => {
    Promise.resolve(book);
  }, []);

  return (
    <Form
      value={{
        author: 'John',
        name: 'HTML/CSS',
        description: 'HTML/CSS with styles',
        imageURL: '',
        publishDate: new Date().getTime(),
      }}
      onSubmit={onSubmit}
    />
  );
};

export default {
  title: 'Practice/components/Form',
  component: Form,
} as Meta;
