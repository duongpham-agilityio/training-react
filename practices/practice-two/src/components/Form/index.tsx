import { MouseEvent, memo, useCallback } from 'react';
import isEqual from 'react-fast-compare';

// Hooks
import { useBookForm, FormData } from 'hooks';

// Components
import { Button, Heading, Input } from 'components/commons';
import BookRecommend from './Recommend';
import ChooseImage from './ChooseImage';

// Styles
import styles from 'components/Form/index.module.css';

export type FormProps = {
  value: FormData;
  title?: string;
  className?: string;
  type?: 'update' | 'create';
  onClose?: () => void;
  onSubmit: (data: FormData) => Promise<void>;
};

const Form = (props: FormProps) => {
  const {
    value: data,
    title = 'Create a new book marker',
    className = '',
    type = 'create',
    onClose,
    onSubmit,
  } = props;
  const {
    value: { author, description, imageURL, name, publishDate, imageName },
    refImage,
    isUpload,
    isSubmit,
    booksRecommended,
    handleSelectRecommended,
    resetRecommended,
    onChange,
    onSubmit: onSubmitOnHook,
  } = useBookForm(data, type, onSubmit);

  const chooseImage = useCallback(() => {
    refImage.current?.click();
  }, [refImage]);

  const resetRecommendedData = useCallback(
    (e: MouseEvent) => {
      e.stopPropagation();

      resetRecommended([]);
    },
    [resetRecommended]
  );

  return (
    <section className={styles.overlay} onClick={onClose}>
      <form
        className={`${styles.form} ${className}`}
        action="#"
        method="POST"
        onSubmit={onSubmitOnHook}
        onClick={resetRecommendedData}
        data-testid="book-form"
      >
        <Heading label={title} className={styles.heading} />

        <section className={styles.content}>
          <div className={styles.formItem}>
            <div className={styles.formField}>
              <label className={styles.label} htmlFor="bookName">
                Book name
              </label>
              <Input
                value={name}
                placeholder="Book name"
                name="name"
                id="bookName"
                onChange={onChange}
              />

              {!!booksRecommended.length && (
                <BookRecommend
                  books={booksRecommended}
                  onSelect={handleSelectRecommended}
                />
              )}
            </div>
            <div className={styles.formField}>
              <label className={styles.label} htmlFor="author">
                Author
              </label>
              <Input
                value={author}
                placeholder="Author"
                name="author"
                onChange={onChange}
              />
            </div>
          </div>

          <div className={styles.formItem}>
            <div className={styles.formField}>
              <label className={styles.label} htmlFor="publishedDate">
                Published date
              </label>
              <Input
                type="date"
                value={publishDate}
                name="publishDate"
                onChange={onChange}
              />
            </div>
            <div className={`${styles.formField} ${styles.upload}`}>
              <div className={styles.uploadItem}>
                <label className={styles.label} htmlFor="imageURL">
                  Upload
                </label>
                <Input
                  ref={refImage}
                  hidden
                  className={styles.input}
                  type="file"
                  name="imageURL"
                  id="imageURL"
                  accept=".png,.jpeg"
                  onChange={onChange}
                />

                <ChooseImage name={imageName} chooseHandler={chooseImage} />
              </div>
              <div className={styles.uploadItem}>
                {imageURL && (
                  <img
                    src={imageURL}
                    alt="image book"
                    className={styles.image}
                  />
                )}
              </div>
            </div>
          </div>

          <div className={styles.formItem}>
            <div className={styles.formField}>
              <label className={styles.label} htmlFor="bookName">
                Description
              </label>
              <Input
                variant="area"
                className={styles.text}
                value={description}
                placeholder="Book name"
                name="description"
                onChange={onChange}
              />
            </div>
          </div>

          <div className={styles.action}>
            <Button
              label="Cancel"
              width="w-lg"
              border="b-lg"
              onClick={onClose}
              disabled={isSubmit}
            />
            <Button
              disabled={isUpload || isSubmit}
              type="submit"
              label="Save"
              variant={isUpload ? 'default' : 'primary'}
              width="w-lg"
              border="b-lg"
            />
          </div>
        </section>
      </form>
    </section>
  );
};

export default memo(Form, isEqual);
