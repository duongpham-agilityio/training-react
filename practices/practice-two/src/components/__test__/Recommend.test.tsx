import { act, fireEvent, render } from '@testing-library/react';

// Types
import { RecommendProps } from 'components/Form/Recommend';

// Components
import BookRecommend from 'components/Form/Recommend';

const setup = (props: RecommendProps) => render(<BookRecommend {...props} />);

describe('BookRecommend', () => {
  const mockProps: RecommendProps = {
    books: [
      {
        name: 'Book 1',
        description: 'Something description',
        imageURL:
          'http://books.google.com/books/content?id=51YEAAAAMBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api',
      },
      {
        name: 'Book 2',
        description: 'Something description',
        imageURL:
          'http://books.google.com/books/content?id=51YEAAAAMBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api',
      },
    ],
    onSelect: jest.fn(),
  };
  it('Match snapshot', () => {
    const { container } = setup(mockProps);

    expect(container).toMatchSnapshot();
  });

  it('Render with props', () => {
    const { getByText } = setup(mockProps);
    const {
      books: [one],
      onSelect,
    } = mockProps;
    const item = getByText(one.name);

    act(() => {
      fireEvent.click(item);
    });

    expect(onSelect).toBeCalled();
  });
});
