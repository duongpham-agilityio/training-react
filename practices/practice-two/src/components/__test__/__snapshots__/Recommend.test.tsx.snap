// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookRecommend Match snapshot 1`] = `
<div>
  <ul>
    <li>
      <p>
        Book 1
      </p>
    </li>
    <li>
      <p>
        Book 2
      </p>
    </li>
  </ul>
</div>
`;
